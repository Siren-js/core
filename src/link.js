import * as coerce from './util/coerce';
import extendWith from './util/extend-with';
import { isArray, isNonNullObject, isString, isUri } from './util/type-guard';

export class Link {
    #rel;
    #href;
    #class;
    #title;
    #type;

    constructor(rel, href, options = {}) {
        if (!isString(rel) && !isArray(rel)) {
            throw new TypeError('Link.rel must be an array of strings');
        }
        if (!isUri(href)) {
            throw new TypeError('Link.href must be a URI');
        }

        const { class: linkClass, title, type, ...extensions } = options ?? {};

        this.rel = rel;
        this.href = href;
        this.class = linkClass;
        this.title = title;
        this.type = type;

        extendWith(this, extensions);
    }

    get rel() {
        return this.#rel;
    }

    set rel(value) {
        this.#rel = coerce.toStringArray(value, this.rel);
    }

    get href() {
        return this.#href;
    }

    set href(value) {
        this.#href = coerce.toUriReference(value, this.href);
    }

    get class() {
        return this.#class;
    }

    set class(value) {
        this.#class = coerce.toOptionalStringArray(value, this.class);
    }

    get title() {
        return this.#title;
    }

    set title(value) {
        this.#title = coerce.toOptionalString(value, this.title);
    }

    get type() {
        return this.#type;
    }

    set type(value) {
        this.#type = coerce.toOptionalMediaTypeString(value, this.type);
    }

    toJSON() {
        const {
            rel,
            href,
            class: linkClass,
            title,
            type,
            ...extensions
        } = this;
        return { rel, href, class: linkClass, title, type, ...extensions };
    }

    static isValid(value) {
        return (
            value instanceof Link ||
            (isNonNullObject(value) &&
                (isArray(value.rel) || isString(value.rel)) &&
                isUri(value.href))
        );
    }

    static of(value) {
        if (value instanceof Link) {
            return value;
        }
        const { rel, href, ...rest } = value;
        return new Link(rel, href, rest);
    }
}
