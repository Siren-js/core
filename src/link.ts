/**
 * Copyright (C) 2020  Dillon Redding
 *
 * See the full notice in the LICENSE file at the top level of the repository.
 */
import * as coerce from './util/coerce';
import Extendable from './util/extendable';
import { isArray, isMediaTypeString, isRecord, isString, isStringArray, isUndefined, isUri } from './util/type-guards';

/**
 * Creates a `Link` object
 * @param rel List of link relation types
 * @param href URI of the linked resource
 * @param options Object containing optional link members (e.g., `title`, `type`) and extensions
 * @throws {TypeError} if `href` is an invalid URI
 */
export function link(rel: readonly string[] | string, href: string | URL, options: Partial<Link> = {}): Link {
    const { 'class': linkClass, title, type, ...extensions } = options;
    let _rel = coerce.toStringArray(rel);
    let _href = coerce.toUriString(href);
    let _class = coerce.toOptionalStringArray(linkClass);
    let _title = coerce.toOptionalString(title);
    let _type = coerce.toMediaTypeString(type);
    return {
        get rel() { return _rel; },
        set rel(value) {
            _rel = coerce.toStringArray(value);
        },

        get href() { return _href; },
        set href(value) {
            _href = coerce.toUriString(value);
        },

        get class() { return _class; },
        set class(value) {
            _class = coerce.toOptionalStringArray(value);
        },

        get title() { return _title; },
        set title(value) {
            _title = coerce.toOptionalString(value);
        },

        get type() { return _type; },
        set type(value) {
            _type = coerce.toMediaTypeString(value);
        },

        ...extensions
    };
}

/**
 * Describes a Siren [link](https://github.com/kevinswiber/siren#links-1), which
 * represents a navigational transition.
 */
export interface Link extends Extendable {
    /**
     * Describes aspects of the link based on the current representation.
     */
    class?: readonly string[];

    /**
     * The URI of the linked resource.
     */
    href: string;

    /**
     * Defines the relationship of the link to its entity, per
     * [RFC 8288](http://tools.ietf.org/html/rfc8288).
     */
    rel: readonly string[];

    /**
     * Text describing the nature of a link.
     */
    title?: string;

    /**
     * Defines the media type of the linked resource, per
     * [RFC 8288](http://tools.ietf.org/html/rfc8288).
     */
    type?: string;
}

/**
 * Determines whether `value` is a valid `Link`.
 */
export function isLink(value: unknown): value is Link {
    return isParsableLink(value) &&
        isStringArray(value.rel) &&
        isString(value.href) &&
        (isUndefined(value.class) || isStringArray(value.class)) &&
        (isUndefined(value.title) || isString(value.title)) &&
        (isUndefined(value.type) || isMediaTypeString(value.type));
}

/**
 * Determines whether `value` is a `ParsableLink`, meaning it can be
 * destructured and passed to `link()`.
 */
export function isParsableLink(value: unknown): value is ParsableLink {
    return isRecord(value) && !isUndefined(value.rel) && isUri(value.href);
}

export interface ParsableLink extends Extendable {
    rel: unknown;
    href: string | URL;
}
