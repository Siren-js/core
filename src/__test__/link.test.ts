import * as Siren from '../link';

const href = 'http://example.com';

describe('link function', () => {
    it('should coerce rel parameter to string array', () => {
        const cases = [
            [['self'], ['self']],
            [['collection', 'up'], ['collection', 'up']],
            [[], []],
            [undefined, []],
            [null, []],
            [true, ['true']],
            [42, ['42']],
            ['foo', ['foo']],
            [[true, 42, 'foo'], ['true', '42', 'foo']],
            [{}, ['[object Object]']]
        ];

        cases.forEach(([value, expected]) => {
            const link = Siren.link(value as string, href);
            expect(link.rel).toEqual(expected);
        });
    });

    describe('href parameter', () => {
        it('should coerce to URI string', () => {
            const cases = [
                [href, href],
                ['/orders', '/orders'],
                [new URL(href), `${href}/`],
                [true, '/true'],
                [42, '/42'],
                ['foo', '/foo'],
                [[true, 42, 'foo'], '/true,42,foo'],
                [{}, '/[object%20Object]']
            ];

            cases.forEach(([value, expected]) => {
                const action = Siren.link('create', value as string);
                expect(action.href).toEqual(expected);
            });
        });

        it('should reject invalid URI', () => {
            [null, undefined, 'http://\uFFFF.com'].forEach(value => {
                expect(() => Siren.link('create', value as string)).toThrow(TypeError);
            });
        });
    });

    describe('optional parameter', () => {
        it('should set optional members to undefined when empty', () => {
            const rel = ['self'];

            const link = Siren.link(rel, href);

            expect(link.rel).toEqual(rel);
            expect(link.href).toEqual(href);
            expect(link.class).toBeUndefined();
            expect(link.title).toBeUndefined();
            expect(link.type).toBeUndefined();
        });

        it('should coerce class option to string array or undefined', () => {
            const cases = [
                [['order'], ['order']],
                [['customer', 'info'], ['customer', 'info']],
                [[], []],
                [undefined, undefined],
                [null, undefined],
                [true, ['true']],
                [42, ['42']],
                ['foo', ['foo']],
                [[true, 42, 'foo'], ['true', '42', 'foo']],
                [{}, ['[object Object]']]
            ];

            cases.forEach(([value, expected]) => {
                const link = Siren.link(['self'], href, { class: value as string });
                expect(link.class).toEqual(expected);
            });
        });

        it('should coerce title option to string or undefined', () => {
            const cases = [
                ['foo', 'foo'],
                ['foo bar', 'foo bar'],
                ['', ''],
                [undefined, undefined],
                [null, undefined],
                [true, 'true'],
                [42, '42'],
                [[true, 42, 'foo'], 'true,42,foo'],
                [{}, '[object Object]']
            ];

            cases.forEach(([value, expected]) => {
                const link = Siren.link(['self'], href, { title: value as string });
                expect(link.title).toEqual(expected);
            });
        });

        describe('type option', () => {
            it('should accept any valid media type string', () => {
                ['application/json', 'text/html', 'image/png'].forEach(type => {
                    const link = Siren.link(['self'], href, { type });
                    expect(link.type).toEqual(type);
                });
            });

            it('should coerce invalid media type string to undefined', () => {
                [undefined, null, true, 42, '', 'foo', [true, 42, 'foo'], {}].forEach(value => {
                    const link = Siren.link(['self'], href, { type: value as string });
                    expect(link.type).toBeUndefined();
                });
            });
        });

        it('should accept extensions', () => {
            const hreflang = 'en-US';

            const link = Siren.link(['self'], href, { hreflang });

            expect(link.hreflang).toEqual(hreflang);
        });
    });

    it('should create immutable object', () => {
        const link = Siren.link(['self'], href, {
            class: ['order'],
            title: 'Order',
            type: 'application/vnd.siren+json',
            hreflang: 'en-US'
        });

        const mutations: ((link: Siren.Link) => void)[] = [
            link => (link.rel as string[]) = ['foo'],
            link => (link.rel as string[])[0] = 'foo',
            link => (link.href as string) = 'foo',
            link => (link.class as string[]) = ['foo'],
            link => (link.class as string[])[0] = 'foo',
            link => (link.title as string) = 'foo',
            link => (link.type as string) = 'foo',
            link => (link.hreflang as string) = 'foo'
        ];

        mutations.forEach(mutate => {
            expect(() => mutate(link)).toThrow(TypeError);
        });
    });
});

describe('link type guard', () => {
    it('should return true for valid links', () => {
        const values: Siren.Link[] = [
            // minimal, absolute URI
            { rel: ['self'], href },
            // minimal, relative URI
            { rel: ['self'], href: '/orders' },
            // kitchen sink
            {
                rel: ['self'],
                href,
                class: ['order'],
                title: 'Order 66',
                type: 'application/xhtml+xml'
            },
            // with extension
            { rel: ['self'], href, hreflang: 'en-US' }
        ];

        values.forEach(value => {
            expect(Siren.isLink(value)).toEqual(true);
        });
    });

    it('should return false for invalid links', () => {
        const values = [
            // non-records
            undefined, null, true, 42, 'foo', [],
            // empty record
            {},
            // invalid rel
            { rel: 'self', href },
            // invalid href
            { rel: ['self'], href: 'http://\uFFFF.com' },
            { rel: ['self'], href: {} },
            // invalid class
            { rel: ['self'], href, class: 'order' },
            // invalid title
            { rel: ['self'], href, title: 42 },
            // invalid type
            { rel: ['self'], href, type: 'order' }
        ];

        values.forEach(value => {
            expect(Siren.isLink(value)).toEqual(false);
        });
    });
});
