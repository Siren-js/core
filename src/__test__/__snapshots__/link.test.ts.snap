// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link serialization 1`] = `
"{
  \\"rel\\": [
    \\"self\\"
  ],
  \\"href\\": \\"http://example.com\\",
  \\"class\\": [
    \\"home\\"
  ],
  \\"title\\": \\"Home Page\\",
  \\"type\\": \\"text/html\\",
  \\"hreflang\\": \\"en-US\\"
}"
`;
